### ISSUE TITLE
Eager renaming

### ISSUE THREAD

**Issue by Janiczek** [jwoudenberg/elm-pair#8]
Continuation of
* https://twitter.com/janiczek/status/1488279032911339520
* https://twitter.com/jasperwoudnberg/status/1488602216386736130
* and the commit 258251d48b08d07eca52d13923353bc2d4fa128d

This still triggers in the "change which constructor I'm casing on" situation on `release-6` (c3dec4055118f404621dd208d30b2803b89351ee):

(`0:16 - 0:20` in particular)

https://user-images.githubusercontent.com/149425/152044691-740abe72-5cf4-4ab5-bba0-f305a9c59363.mp4

**Comment by jwoudenberg** [comment]
Huh, okay, did not expect that! Thank you again for the clear bug report, I'm going to take a look at this tomorrow!

**Comment by jwoudenberg** [comment]
Ohh, I guess this makes sense. I changed the logic so renames into variables names that exist elsewhere in the code don't take place. But we're not doing a single rename, but many small ones (on a fast enough machine we rename a letter at a time). The final rename from `Identifie` to `Identifier` falls foul of the new logic and gets aborted, which just looks buggy. I guess my last change didn't really make things better.

Going to think a bit about this and if I can't come up with anything else, then go with the approach you previously proposed on twitter and only accept renames if one changes the source definition.

**Comment by Janiczek** [comment]
In my mind, the more I think about it, the more I'm of the opinion that in case of changing a `case..of` pattern like this:
```elm
type X = String ()

fn1 x =
  case x of
    String a -> a 

fn2 x =
  case x of
    String a -> a 
  --^ cursor is here
```
changing into
```elm
type X = String ()

fn1 x =
  case x of
    String a -> a 

fn2 x =
  case x of
      a -> a 
  --^ cursor is here
```
it should _not_ classify the edit as a "please rename this constructor everywhere, including in the original type" action (no matter the number of occurences of the `String` constructor). But maybe that's just my particular style of editing and other people always start a new line and write it from scratch instead of copypasting a similar piece of code and editing it. I don't know... 🤔

**Comment by jwoudenberg** [comment]
Yeah, I totally do the copy-paste thing too. I think using Elm-pair is only going to be a pleasant experience when it guesses intent correctly almost without fail. I think what you're doing in the demo video is totally reasonable and Elm-pair doesn't guess intent correctly, so I definitely agree current behavior is a bug. Also your example uses constructors, but I can also imagine situations where we change a function call because we want to call a different function, which Elm-pair would currently also interpret as you wanting to rename the original function, so the problem is broader.

I think not needing to be at the callsite to change a name can be nice, because it saves you a 'go to definition' (either manually or assisted), but don't see a way to have that while not running into this bug, so I'm going to take the approach you suggested in tweet and only rename when you change the name at the definition site.

**Comment by jwoudenberg** [comment]
This is out as part of release 7!

**Comment by Janiczek** [comment]
Awesome! 🙂 Trying to break this some more, on release 7:

I'm trying to change `Number _ ->` into `String _ ->`. Only the currently edited line changes (good!) but as soon as I finish `String`, all the other `String`s change to `String2`. I didn't expect that. WDYT?

https://user-images.githubusercontent.com/149425/152218149-28b18fe2-f572-4e64-a958-0c3371f27380.mp4

**Comment by jwoudenberg** [comment]
Ugh, I should have caught this in my own testing. Thank you for your patience with this issue! I'm now pushing Release 8 with a fix.

**Comment by Janiczek** [comment]
TL;DR: seems like the Darwin binary for Release 8 is broken.

The download script wasn't able to download a working binary for Release 8 for macOS 🤔 I tried two times from scratch, each time starting with the contents of the neovim-plugin directory from [this revision](87bafb92aa7be985bae7de52414656710857f886) of your repo. Also tried to download the [darwin](https://github.com/jwoudenberg/elm-pair/releases/download/release-8/elm-pair_x86_64-darwin.tar.gz) asset from your [releases page](https://github.com/jwoudenberg/elm-pair/releases/tag/release-8) manually, with the same result.

<img width="691" alt="Screenshot 2022-02-03 at 19 10 57" src="https://user-images.githubusercontent.com/149425/152403171-bf09c4df-b2f6-4cb8-8015-51f224b3adb3.png">

(It _was_ suspicious that the .tar.gz compressed so well! 😁 )

![Screenshot 2022-02-03 at 19 16 26](https://user-images.githubusercontent.com/149425/152403948-6dc52ec0-3ea7-4170-8558-0310831c0231.png)

**Comment by Janiczek** [comment]
> Thank you for your patience with this issue!

To the contrary, thank _you_ for putting up with my neverending issues! 😄 I'm happy to go through this process, and I'm glad you're improving the tool because I'm having an absolute blast with elm-pair so far!

**Comment by jwoudenberg** [comment]
Thank you! Unclear why the github release pipeline didn't work. Rerunning it has the same result. I uploaded a manually compiled binary for now, that ought to do the trick. Thank you for the report!

**Comment by Janiczek** [comment]
I still had some (different) issues with the updated binary:
<img width="649" alt="Screenshot 2022-02-04 at 0 13 06" src="https://user-images.githubusercontent.com/149425/152444900-3f17ef15-0682-4281-ab99-0a7a29357206.png">

Anyways I compiled elm-pair myself for now (doing `cargo build --release` in `nix-shell -p cargo rustc libiconv darwin.apple_sdk.frameworks.CoreServices`), and now it works again.

Trying out the `Number _ ->` -> `String _ ->` scenario, I can't find any immediate issues with it anymore! :tada:

Thanks for fixing this :)

**Comment by jwoudenberg** [comment]
Ah yeah, I guess that's the risk of building on my Nix machine: it links to dynamic C dependencies from the Nix store making the binaries not work on other machines. Thanks for letting me know, I'm going to look into fixing the release script today. And glad you got it to work in the meanwhile!

I'll close this issue and create a new one for the binary problem.
